### ISSUE TITLE
Issue with const enums and TypeScript

### ISSUE THREAD

**Issue by rajyraman** [parcel-bundler/parcel#3624]
I noticed that when I compile a TypeScript file with const enums, parcel's compiled JS looks incorrect compared to commandline tsc compile. For e.g. I have this simple TypeScript code

```javascript
declare const enum Colour {
  Green,
  Blue
}
const myColour = Colour.Green;
console.log(`Green: ${myColour === Colour.Green}`);
```
Below is tsc's output which is correct.
```javascript
var myColour = 0 /* Green */;
console.log("Green: " + (myColour === 0 /* Green */));
```
Below is parcel's output, where the enum is not replaced with a integer value.
```javascript
var myColour = Colour.Green;
console.log("Green: " + (myColour === Colour.Green));
```
I tried with the _preserveConstEnums_ set to false in tsconfig, and it did not make any difference. Is there a reason why Parcel is not replacing enums with int?

**Comment by adhrinae** [comment]
As far as I know, now `parcel-bundler` uses Babal + Babel Typescript plugin for in order to compile typescript codes.

Then there are few limitations on `@babel/plugin-transform-typescript`. [Here is the link of the documentation](@babel/plugin-transform-typescript
).

There are noticeable differences from direct typescript compilation:
* no `namespace` support
* no `const enum` support

**Comment by rajyraman** [comment]
@adhrinae - I went through the source, and based on my understanding, it looks like it is running local typescript compiler.

https://github.com/parcel-bundler/parcel/blob/04643034543330b4cbf7168c6cae64e4d7c20038/packages/core/parcel-bundler/src/assets/TypeScriptAsset.js#L16-L50

**Comment by adhrinae** [comment]
You are right. I forgot about several contexts since I don't use parcel-bundler anymore.

Those errors which you are getting through had happened after version `1.10.0` (Updated to Babel 7).  
Looks like it doesn't have any relationship between updating Babel version and Typescript, however, that brought problems for using Typescript with the parcel.
Sorry for the misguidance.

**Comment by rajyraman** [comment]
@adhrinae - No worries. I used webpack and ts-loader for cross verification and it produces the expected output.

**Comment by garthenweb** [comment]
Looks like an issue with the tsconf in the first place to me. Do you use a custom configuration or the default parcel config? And is webpack using the same config file and typescript version?
Also, out of interest: Is there a reason why the enum is declared with `declare const enum` instead of just `enum`. I assume that `declare` causes the problem.

**Comment by rajyraman** [comment]
I compiled the same code with the same tsconfig across tsc, webpack and parcel. Removing declare does not make any difference.

**Comment by mischnic** [comment]
>  it looks like it is running local typescript compiler.

Correct, only Parcel 2 will switch to using Babel by default here.

---

Parcel calls `transpileModule`, 
```js
const typescript = require('typescript');
const INPUT = `declare const enum Colour {
  Green,
  Blue
}
const myColour = Colour.Green;
console.log(\`Green: $\{myColour === Colour.Green}\`);`;

let transpilerOptions = {};

console.log(typescript.transpileModule(INPUT, transpilerOptions).outputText);
/*
var myColour = Colour.Green;
console.log("Green: " + (myColour === Colour.Green));
*/
```

the output does differ from the TS playground output (or your webpack output): http://www.typescriptlang.org/play/index.html?ssl=1&ssc=9&pln=1&pc=1#code/CYUwxgNghgTiAEYD2A7AzgF3iFBXAtvAMJIRK4zwDeAUPPAOJw4A0d8AQhLiDQL41k6LPgCeJMhXgBeYqXIwAdExA4A3INRpSIRWQDmACgAGKnAC54AEipiJCmdNn2Ky5ij7GAlGqA

This seems to be a (intentional) limitation of `transpileModule`:
https://github.com/Microsoft/TypeScript/issues/5243#issuecomment-147766222
> `const` enums are treated as regular enums (even for locally defined `const` enums)

**Comment by rajyraman** [comment]
But the enum declaration and usage are in the same file. In the linked issue the enum is in a different declaration file.

**Comment by mischnic** [comment]
> In the linked issue the enum is in a different declaration file.

as the quote says:

> const enums are treated as regular enums (**even for locally defined** const enums)
